### PR TITLE
Add ability to enable a graphql API endpoint in the Admin UI. Add example for this feature in examples-next

### PR DESCRIPTION
**Pull request by rohan-deshpande**
### Added 

* Config option for toggling the graphql api endpoint in the admin ui
* Template for graphql api endpoint to be generated for the admin ui

### Changed 

* The ApolloServer is no longer added to express if the new config option is enabled as the API is made available through a next-js API route

### CODE REVIEW

File: packages-next/admin-ui/src/templates/index.ts [hunk 1]
```diff
@@ -37,4 +38,5 @@ export const writeAdminFiles = (
     ({ adminUILabels: { path }, key }) =>
       ({ mode: 'write', src: itemTemplate(key), outputPath: `pages/${path}/[id].js` } as const)
   ),
+  { mode: 'write', src: apiTemplate, outputPath: 'pages/api/graphql.js' },
```
Comment (mitchellhamilton) [inline packages-next/admin-ui/src/templates/index.ts:41]: Should be conditional based on the option.

Comment (rohan-deshpande) [reply packages-next/admin-ui/src/templates/index.ts:41]: Yah I tried this but typescript was very angry at me. Will try again.

File: packages-next/admin-ui/src/system/createAdminUIServer.ts [hunk 2]
```diff
@@ -11,14 +10,16 @@ export const createAdminUIServer = async (
   projectAdminPath: string,
   sessionStrategy?: SessionStrategy<any>
 ) => {
+  const thing = 'n' + 'ext';
```
Comment (mitchellhamilton) [inline packages-next/admin-ui/src/system/createAdminUIServer.ts:13]: Probs worth adding a comment above this and the other one to say this is to stop webpack from bundling next inside neext.

Comment (rohan-deshpande) [reply packages-next/admin-ui/src/system/createAdminUIServer.ts:13]: Is the string concat actually needed here? Is that thing that will _trick_ webpack?

File: packages-next/admin-ui/src/templates/index.ts (hunk 1 above)
Comment (mitchellhamilton) [reply packages-next/admin-ui/src/templates/index.ts:41]: The answer to TypeScript complaining will probably be to write `as const` after `mode: 'write'`

File: packages-next/admin-ui/src/system/createAdminUIServer.ts (hunk 2 above)
Comment (mitchellhamilton) [reply packages-next/admin-ui/src/system/createAdminUIServer.ts:13]: Don't think so since it works on the other one without the string concat. Having the string in a variable and then calling require with it is def needed though.

**Comment by mitchellhamilton** [comment]
You're gonna have to run `yarn manypkg fix` to fix CI

File: packages-next/admin-ui/src/system/createAdminUIServer.ts (hunk 2 above)
Comment (rohan-deshpande) [reply packages-next/admin-ui/src/system/createAdminUIServer.ts:13]: Yah cool, will just put it in a var and dump the concat then, cheers!

**Review by mitchellhamilton — COMMENTED** [review]
Looks fine to me(other than it needs changesets and CI needs to pass) now but I feel like @timleslie will have opinions on this

File: packages-next/keystone/src/lib/createKeystone.ts [hunk 3]
```diff
@@ -25,7 +25,7 @@ export function createKeystone(config: KeystoneConfig, dotKeystonePath: string,
     adapter = new MongooseAdapter({ mongoUri: db.url, ...db.mongooseOptions });
   } else if (db.adapter === 'prisma_postgresql') {
     adapter = new PrismaAdapter({
-      getPrismaPath: () => path.join(dotKeystonePath, 'prisma'),
+      getPrismaPath: () => '.keystone/prisma',
```
Comment (timleslie) [inline packages-next/keystone/src/lib/createKeystone.ts:28]:
```suggestion
      getPrismaPath: () => path.join(dotKeystonePath, 'prisma'),

```

File: packages-next/types/src/config/index.ts [hunk 4]
```diff
@@ -85,6 +85,8 @@ export type AdminUIConfig = {
     isValidSession: boolean;
     createContext: CreateContext;
   }) => MaybePromise<{ kind: 'redirect'; to: string } | void>;
+  /** Enables nextjs graphql api route mode */
+  enableGraphQlApiEndpoint?: boolean;
```
Comment (timleslie) [inline packages-next/types/src/config/index.ts:89]: This will need to be documented in the  API docs in the "Advanced configuration:" section of `docs-next/pages/apis/config.mdx`.

File: packages-next/admin-ui/src/templates/api.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+export const apiTemplate = `
```
Comment (timleslie) [inline packages-next/admin-ui/src/templates/api.ts:1]: This code is a long way away from the very similar code in `createExpressServer`.

I wonder if there's a way we can bring those two things closer together? 🤔 

Also, since this isn't paramterised, should we put in in `static` and use `copy` operation rather than a `write`?

**Review by timleslie — CHANGES_REQUESTED** [review]
A few minor nits and thoughts to follow up on.

File: packages-next/types/src/config/index.ts (hunk 4 above)
Comment (JedWatson) [reply packages-next/types/src/config/index.ts:89]: oof, is that how we capitalise `enableGraphQlApiEndpoint`?

Took a look at the types and we're not super consistent here, but an equivalent config option is `extendGraphqlSchema` so I'm pretty sure we decided that "Graphql" is just a word in terms of camelcasing

Also I get why this is nested under `ui` but it's ... weird. I misunderstood what this option meant on the first read. We should come back to that later, in the meantime my preferred name for this is

```
  /** Enables nextjs graphql api route mode */
  addNextjsGraphqlEndpoint?: boolean;
```

Basically, tells the ui to "add a Next.js GraphQL Endpoint", and then when we're creating an express server, if that's the case we don't also add an express handler for that.

**Comment by JedWatson** [comment]
Actually, carrying forward with my config nits, it's slightly outside the scope of this PR but maybe we can make a quick call.

This config option really _should_ be added now, but also, we need to loop back and think more about how the whole API works in terms of where we're generating the graphql endpoint (and how that might change between dev, prod, and different deployment targets)

So I propose we add a top-level `experimental` grab bag of config that lets us say "this is a thing we're trying out"

The change to this PR would be moving the config key `enableGraphQlApiEndpoint` here:

```
type KeystoneConfig = {
  /* ... */
  experimental?: {
    /** Adds an API route to the Admin UI to serve the graphql endpoint, and skips adding express middleware for it */
    useNextjsRouteForGraphqlEndpoint?: boolean;
  }
}
```

I think that resolves all of my concerns about getting this shipped; it's an experimental feature, and also I'm not sure where the config should live yet.

File: packages-next/keystone/src/lib/createKeystone.ts (hunk 3 above)
Comment (mitchellhamilton) [reply packages-next/keystone/src/lib/createKeystone.ts:28]: not resolving this path is important so that adapter-prisma will do it because `@keystone-next/keystone` will be bundled in the next app(but `@keystonejs` packages won't be, re: why just `@keystonejs`? because it's an easy line to draw because it's all backend code) and webpack will essentially mock `process.cwd()` to return `"/"` so without this change, the prisma adapter would try to write to `/.keystone/prisma`(as in at the root of the file system).

File: packages-next/admin-ui/src/templates/api.ts (hunk 5 above)
Comment (rohan-deshpande) [reply packages-next/admin-ui/src/templates/api.ts:1]: > This code is a long way away from the very similar code in createExpressServer.

Hmm but there does seem to be a fairly heavy codependency between `admin-ui` and `system` it's already there as `createAdminUIServer` is imported into `createExpressServer` 

> Also, since this isn't paramterised, should we put in in static and use copy operation rather than a write?

Easily done!

**Comment by rohan-deshpande** [comment]
- [ ] Add experimental config and put the new config option in there
- [ ] Put the graphql api route page template inside of `static` and use `copy` instead of `write`

File: packages-next/types/src/config/index.ts (hunk 4 above)
Comment (rohan-deshpande) [reply packages-next/types/src/config/index.ts]: Just saw this comment, okay can easily change this

Comment (rohan-deshpande) [reply packages-next/types/src/config/index.ts]: I think enable is better for boolean values though as a key

Comment (rohan-deshpande) [reply packages-next/types/src/config/index.ts]: also do we need to document stuff under `experimental` _right now_?

File: packages-next/admin-ui/src/templates/api.ts (hunk 5 above)
Comment (rohan-deshpande) [reply packages-next/admin-ui/src/templates/api.ts:1]: On second thought I think this should live in templates... I can't see a `static` dir in there but it's definitely a template even if it's not dynamic currently. I think moving it away from there might be a bit weird. 

Lemme know if I've misunderstood something.

File: packages-next/keystone/src/lib/createKeystone.ts (hunk 3 above)
Comment (JedWatson) [reply packages-next/keystone/src/lib/createKeystone.ts:28]: @mitchellhamilton do you mean we _should_ merge @timleslie's suggestion? or not?

Comment (mitchellhamilton) [reply packages-next/keystone/src/lib/createKeystone.ts:28]: I mean that we very much _shouldn't_

File: docs-next/pages/apis/access-control.mdx [hunk 6]
```diff
@@ -12,9 +12,15 @@ export default config({
   lists: createSchema({
     ListKey: list({
       fields: {
-        fieldName: text({ access: { /* ... */ }, }),
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:15]: I'm not sure why these got changed, looks like our prettier ignore might not be configured correctly? or this is maybe something that happened locally

Because this shows up w/ formatting in the docs though, we'll want to roll all the mdx changes back in this PR

@gwyneplaine do you know if we've made any changes to the repo config that would have caused this to happen?

Comment (gwyneplaine) [reply docs-next/pages/apis/access-control.mdx:15]: @JedWatson not aware of any prettier config changes, however these single-line formatted comments were manual commits via suggestion iirc.

I honestly i think we should stop fighting prettier at this point cc @timleslie

Comment (JedWatson) [reply docs-next/pages/apis/access-control.mdx:15]: @gwyneplaine agree on not fighting prettier generally

In this case I'd rather we tell it to ignore `mdx` files entirely if we can't just tell it to ignore `tsx` in `mdx`

This formatting ends up in the docs, and all the extra line breaks hurt readability much more than the benefit having prettier runs on our examples adds, at least at this point imo

Comment (rohan-deshpande) [reply docs-next/pages/apis/access-control.mdx:15]: @JedWatson how do you want me to resolve? Just checkout the MDX files from origin/master and push back up?

File: packages-next/admin-ui/src/system/buildAdminUI.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
 export async function buildAdminUI(projectAdminPath: string) {
   // importing next/dist/build is quite expensive so we're requiring it lazily
-  const build = (require('next/dist/build') as typeof import('next/dist/build')).default;
+  const next = 'next/dist/build';
```
Comment (mitchellhamilton) [inline packages-next/admin-ui/src/system/buildAdminUI.ts:3]:
```suggestion
  /** We do this to stop webpack from bundling next inside of next */
  const next = 'next/dist/build';
```

File: .changeset/slimy-elephants-hear.md [hunk 8]
```diff
@@ -0,0 +1,8 @@
+---
+'keystone-next-app': patch
+'@keystone-next/admin-ui': patch
+'@keystone-next/keystone': patch
+'@keystone-next/types': patch
+---
+
+Added graphql api route example for admin UI
```
Comment (mitchellhamilton) [inline .changeset/slimy-elephants-hear.md:8]: You don't need a changeset for `keystone-next-app` since it's not published but for the other packages, you should provide a more descriptive summary since seeing `Added graphql api route example for admin UI` in the changelog for any of the other packages wouldn't really make sense

File: docs-next/pages/apis/access-control.mdx (hunk 6 above)
Comment (rohan-deshpande) [reply docs-next/pages/apis/access-control.mdx]: Reverted all to `origin/master`

File: .changeset/slimy-elephants-hear.md (hunk 8 above)
Comment (rohan-deshpande) [reply .changeset/slimy-elephants-hear.md]: Done!

**Review by timleslie — APPROVED** [review]
This all looks good to me 👍
